[ghc] add automation

ulidtko opened the pull request:
Followup to https://github.com/endoflife-date/endoflife.date/pull/6287, adds integration for the thing to keep rolling more-or-less autonomously (without regular mandatory attention).

A new script within existing framework.

* Parses the source table from wiki — to gather the "EOL" / "Not recommended for use" signals
* Fills out `releases`
* Fixes `versions` dates to come from wiki tables, rather than git tags (formal version announcements sometimes precede git tag, sometimes lag behind).

I understood that

```diff
--- i/products/ghc.md
+++ w/products/ghc.md
@@ -22,6 +22,7 @@ auto:
   - git: https://gitlab.haskell.org/ghc/ghc.git
     regex: ^ghc-(?P<major>\d+)[.](?P<minor>\d+)[.](?P<patch>\d+)-release$
     template: '{{major}}.{{minor}}.{{patch}}'
+  - custom: ghc-wiki
 
 releases:
```

plus a submodule bump (maybe?), will be needed as another follow-up in the main repo.

:heavy_check_mark:  Tested locally. (Thence the diff in generated json)

ulidtko commented:
@captn3m0 @marcwrobel please review, if/when time permits

marcwrobel commented:
Hi @ulidtko, checked the script:

- There is no need to parse versions IMO as it's already done based on tags on https://github.com/endoflife-date/endoflife.date/blob/eb24d6b0801128b3c84df0505a844ce944576a6b/products/ghc.md?plain=1#L22 (they are not exactly the same, but close enough ?).
- For releases the use of label such as "not recommended for use" is not allowed : it should be either a date or a boolean.
- Still for releases there is no need to set latest version : it is already automatically kept up-to-date based on the versions.

So I think only the releases eoas / eol status should be kept overall. WDYT ?

The submodule bump is automatic (once a day), no need to do anything there.

ulidtko commented:
Hi @marcwrobel, thanks for review.

Re: git tags VS parse from wiki: I thought of that; PR description
> Fixes versions dates to come from wiki tables, rather than git tags (formal version announcements sometimes precede git tag, sometimes lag behind).

— there's been occasionally large-ish gap between git tag & official release announce (~weeks, both ways, early and late).

So I coded the script to give preference to the wiki. IMHO the added precision improves credibility of data on endoflife.date.

> For releases the use of label such as "not recommended for use" is not allowed : it should be either a date or a boolean.

Yes; those *labels* for `eolColumn` and `eoasColumn` were settled in https://github.com/endoflife-date/endoflife.date/pull/6287 — see discussion there.

The *values* in columns themselves are in-line with other products' data; no funny business there. Here's a sample of script output:

```json
    "9.8": {
      "name": "9.8",
      "eol": false,
      "eoas": true,
      "latest": "9.8.4",
      "latestReleaseDate": "2024-12-02"
    },
```

— dates & booleans, exactly as you say.

> Still for releases there is no need to set latest version : it is already automatically kept up-to-date based on the versions.

I didn't see that in testing.

Furthermore, giving preference to wiki data affords this *point of manual override* by humans: suppose the following possible scenario.
1. GHC team tags a version X.Y.99 in git (to get pre-release builds out of CI), and sends it for in-field testing — before formally announcing it as released. (which includes corresponding wiki edits)
2. endoflife.date picks up that tag — fully-automatically — and marks as latest.
3. Testing finds a critical bug. Version is retracted / disqualified for official release, and returned back to bugfix.
4. Now, notice we got inconsistency:
   * version X.Y.99 never appeared in announcements, neither is shown as the latest on wiki;
   * endoflife.date shows it as latest, regardless.

Consider that git tags are essentially immutable; it's easier for release-engineering comms to happen on the project-wiki (& mail lists, but let's not parse that) than in git tags.

That's roughly why I give strong preference to the hand-edited version-status-table on wiki — because it exists purely for communication intent. On the other hand, git tags tend to have technically-motivated ties into CI processes; at least in GHC, it often holds, but strictly speaking, git tag existence doesn't immediately imply version release.

If you want, I can incorporate some of this motivation & insight into the script itself, in comments.

ulidtko commented:
Ping @marcwrobel @captn3m0,

9.12.2 got released yesterday — and it's awesome to see that the project's built-in `git tags` automation has already picked it up! :rocket:  https://endoflife.date/ghc

That's truly great. Well done! **However**, I do see a problem in fact occurring...

I saw it coming — and it's exactly why I wrote this script in the first place, 3 months ago.

The problem: the **EOL-status column lost sync**:

![image](https://github.com/user-attachments/assets/df3978c3-59a1-4606-857a-6c50c95f1f4f)

Per the upstream wiki, 9.2 and 9.4 are already **"Not recommended for use"**  (which is basically, GHC's wording for EOL):

![image](https://github.com/user-attachments/assets/6e9ded80-9190-4dbd-8f20-03ab437befc7)

----------

I've rebased & retested the script, fixed error handling, fixed wiki parsing & adapted it to tweaks in wiki format, expanded in-source docs.

The updates in `releases/ghc.json` — are from me test-running the script. Its not required for me to commit them, I understand, but just to make the PR easier to review.

Please check again.

ulidtko commented on releases/ghc.json:
```diff
@@ -1,17 +1,78 @@
+      "latestReleaseDate": "2023-11-10"
+    },
+    "9.2": {
+      "name": "9.2",
+      "eol": true,
```
This `"eol": true` here, likewise for 9.4 — comes from parsing the wiki table.

Which this `ghc-wiki.py` script I'm adding does.

That's the whole point of it... To keep the EOL column in sync.

marcwrobel commented on src/ghc-wiki.py:
```diff
@@ -0,0 +1,103 @@
+            r.set_eol("not recommended for use" in status or ":red_circle:" in status)
+            #-- eoasColumn label is "Further releases planned"
+            r.set_eoas(any(keyword in nextplanned for keyword in ("None",  "N/A")))
+            #-- Initial release in major series is x.y.1
+            r.set_field('latest', mostrecent)
+            if not mostrecent.endswith('.1'):
+                latest = product.get_version(mostrecent)
+                if latest:
+                    r.set_field('latestReleaseDate', latest.date())
```
```suggestion
```

The latest version is automatically computed based on versions by https://github.com/endoflife-date/release-data/blob/main/latest.py, there is no need to compute it there as it will be ignored anyway.

marcwrobel replied:
Done. We could also remove the git automation given that all versions are listed on https://gitlab.haskell.org/ghc/ghc/-/wikis/GHC-status. WDYT ?

Note that the comment about initial release in major series being x.y.1 does not seem to matter much : there are no tags for `x.y.0` versions.

marcwrobel reviewed (approved):
LGTM, thanks @ulidtko.

Using beautifulsoup would have been more in line with what's done in other scripts but not worth changing it now.

I will perform a few minor updates, most notably the one I pointed out about latest versions in releases, and merge.

marcwrobel commented:
Prepared https://github.com/endoflife-date/endoflife.date/pull/7060 (to be merged after this is merged).

ulidtko commented:
Yep @marcwrobel, no objections to your additions! :open_hands: 

I'd prefer to keep both auto-methods, the git-tags and this script. They both run pretty fast. The reason being: wiki text is fluid, and editors don't normally consider it to be a data-source, so there's high likelihood that format changes will break the wiki-parsing script at some point; I won't be surprised if that happens. In that regard, git tags are much more robust, and at least the versions info will keep flowing. Good backup.

I inquired the project about proper machine-readable source for the info — https://gitlab.haskell.org/ghc/release/release-metadata/-/issues/1 — but it's not something that exists yet.
